_posts/2021-12-09-keras-preprocessing-layers/keras-preprocessing-layers.Rmd

```diff
@@ -0,0 +1,324 @@
+
+Now we have an idea what these layers do for us, let's focus on the specific case of state-preserving layers.
+
+## Pre-processing layers that keep state
+
+A layer that randomly perturbs images doesn't need to know anything about the data. It just needs to follow a rule: With probability $p$, do $x$. A layer that's supposed to vectorize text, on the other hand, needs to have a lookup table, matching character strings to integers. The same goes for a layer that maps contingent integers to an ordered set. And in both cases, the lookup table needs to be built upfront.
+
+With stateful layers, this information-buildup is triggered by calling `adapt()` on a freshly-created layer instance. For example, here we instantiate and "condition" a layer that maps strings to consecutive integers:
+
+```{r}
+colors <- c("cyan", "turquoise", "celeste");
+
+layer <- layer_string_lookup()
+layer %>% adapt(colors)
+
+```
+
+We can check what's in the lookup table:
+
+```{r}
+layer$get_vocabulary()
+```
+
+    [1] "[UNK]"     "turquoise" "cyan"      "celeste"  
+
+Then, calling the layer will encode the arguments:
+
+```{r}
+layer(c("azure", "cyan"))
+```
+
+    tf.Tensor([0 2], shape=(2,), dtype=int64)
+
+`layer_string_lookup()` works on individual character strings, and consequently, is the transformation adequate for string-valued categorical features. To encode whole sentences (or paragraphs, or any chunks of text) you'd use `layer_text_vectorization()` instead. We'll see how that works in our [second end-to-end example](#example-2-text-vectorization).
+
+## Using pre-processing layers for performance
+
+Above, we said that pre-processing layers could be used in two ways: as part of the model, or as part of the data input pipeline. If these are *layers*, why even allow for the second way?
+
+The main reason is performance. GPUs are great at regular matrix operations, such as those involved in image manipulation and transformations of uniformly-shaped numerical data. Therefore, if you have a GPU to train on, it is preferable to have image processing layers, or layers such as `layer_normalization()`, be part of the model (which is run completely on GPU).
+
+On the other hand, operations involving text, such as `layer_text_vectorization()`, are best executed on the CPU. The same holds if no GPU is available for training. In these cases, you would move the layers to the input pipeline, and strive to benefit from parallel -- on-CPU -- processing. For example, here pre-processing makes use of four parallel CPU threads:
+
+```{r}
+# pseudocode
+
+preprocessing_layer <- ... # instantiate layer
+
+dataset <- dataset %>%
+  dataset_map(function(x, y) list(preprocessing_layer(x), y)) %>%
```

Comment:
@t-kalinowski could you please edit the code so this uses 4 parallel workers (in an elegant way :-))?
thanks!